rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+	// The final GraphQL request to the subgraph
+	Request *core.GraphQLRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The original http request to the subgraph
+	Orignal *http.Request
+	// Logger for the subgraph request
+	Logger *zap.Logger
+}
+
+type SubgraphResponse struct {
+	// The active subgraph
+	Subgraph *core.Subgraph
+	// The original Gateway request
+	GatewayRequest *core.GatewayRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The final GraphQL response from the subgraph
+	Response *core.GraphQLResponse
+	// The original http response from the subgraph
+	Orignal *http.Response
+}
+
+type ModuleHooks interface {
+	// Provision is called when the module is provisioned
+	Provision(ctx *core.ModuleContext) error
+	// Shutdown is called when the module is shutdown
+	Shutdown() error
+	// Module returns the module information and factory function
+	Module() core.ModuleInfo
+}
+
+// SubgraphHooks are called when a subgraph request or response is made.
+// The order is not guaranteed, so the hooks should be idempotent and side-effect free.
+// if state needs to be shared between hooks, it should be stored in the context.
+// We will provide an easy way to share state between hooks.
+type SubgraphHooks interface {
+	// OnRequest is called when a subgraph request is made.
+	// Returning an error will result in a GraphQL error as a response from the subgraph.
+	OnSubgraphRequest(req *core.SubgraphRequest, err error) error
+	// OnResponse is called when a subgraph response is received
+	// Returning an error will result in a GraphQL error as a response from the subgraph.
+	OnSubgraphResponse(res *core.SubgraphResponse, err error) error
+	// OnError is called when an error occurs during the subgraph lifecycle request/response or any other error
+	OnError(err error)
+}
+
+type ApplicationHooks interface {
+	// OnStart is called when the application starts
+	OnAppStart() error
+	// OnStop is called when the application stops
+	OnAppStop() error
+	// OnError is called when an error occurs during the application lifecycle start/stop or any other error
+	OnAppError(err error)
+}
+
+type TelemetryHooks interface {
+	// OnSpanStart is called when a span is created
+	// Returning a function to be called when the span ends.
+	// This can be used to add custom attributes or events to the span.
+	OnSpan(span *trace.Span) func() // Return a function to be called when the span ends
+}
+
+type AuthenticationHooks interface {
+	// OnAuthenticate is called when a gateway request is authenticated
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnAuthenticate(req *core.GatewayRequest, err error) error
+	// OnPostAuthenticate is called after a gateway request is authenticated
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnPostAuthenticate(req *core.GatewayRequest, err error) error
+}
+
+type OperationHooks interface {
```

Comment:
rename to GraphQLOperationHooks?
also: https://github.com/wundergraph/cosmo/pull/1063/files#r1718053098

Reply:
Done